Mention TargetingContext and fix wording in Variants section of readme (#330)
microsoft/FeatureManagement-Dotnet: README.md

When new features are added to an application, there may come a time when a feature has multiple different proposed design options. A common solution for deciding on a design is some form of A/B testing, which involves providing a different version of the feature to different segments of the user base and choosing a version based on user interaction. In this library, this functionality is enabled by representing different configurations of a feature with variants.

> Variants enable a feature flag to become more than a simple on/off flag. A variant represents a value of a feature flag that can be a string, a number, a boolean, or even a configuration object. A feature flag that declares variants should define under what circumstances each variant should be used, which is covered in greater detail in the [Allocating a Feature's Variants](./README.md#allocating-a-features-variants) section.

Reviewer: the reference should be Allocating Variants now.

Reviewer: Given some other section names were updated, it's worth a double check if any links need updated pointing to those.

Author: Ok I'll double check

}
```

> ### Getting Variants

Reviewer: I think it would be better if the code snippet came directly after this part "For each feature, a variant can be retrieved using the IVariantFeatureManager's GetVariantAsync method."

then after the code snippet the extra information was provided.

Reviewer: I was thinking in our doc it might be helpful to mention that .NET makes it easy to bind IconfigurationSection to a type. After all, this is how I expect most people to consume a features variants. After seeing this, I think a second code snippet could be provided. something like below

---

Once a variant is retrieved, the configuration of a variant can be used directly as an `IConfigurationSection` from the variants `Configuration` property. Another option is to bind the configuration to a type using .NET's configuration binding pattern.

```
IConfigurationSection variantConfiguration = variant.Configuration;

MyFeatureSettings settings = variant
    .Configuration
    .Bind<MyFeatureSettings>();
```

Reviewer 2: ``` C#
MyFeatureSettings settings = variantConfiguration.Get<MyFeatureSettings>() ?? new MyFeatureSettings();
```
Or
```
MyFeatureSettings settings = new MyFeatureSettings();
variantConfiguration.Bind(settings);
```

For each feature, a variant can be retrieved using the `IVariantFeatureManager`'s `GetVariantAsync` method. The variant returned is dependent on the
user currently being evaluated, and that information is obtained from an instance of `TargetingContext`. This context can either be passed in when

> calling `GetVariantAsync` or it can be automatically retrieved from an implementation of `ITargetingContextAccessor` if one is registered.

Reviewer: I feel that this should link to the `ITargetingContextAccessor` section for more details.

// Do something with the resulting variant and its configuration
```

> ### Defining a Feature's Variants

Reviewer 2: I think "Defining Variants" would be better since the above section is called "Getting Variants"

Reviewer: I like the suggestion.

Reviewer: Same with `Allocating Variants`

Each variant has two properties: a name and a configuration. The name is used to refer to a specific variant, and the configuration is the value of that variant. The configuration can be set using either the `ConfigurationReference` or `ConfigurationValue` properties. `ConfigurationReference` is a string path that references a section of the current configuration that contains the feature flag declaration. `ConfigurationValue` is an inline configuration that can be a string, number, boolean, or configuration object. If both are specified, `ConfigurationValue` is used. If neither are specified, the returned variant's `Configuration` property will be null.

A list of all possible variants is defined for each feature under the `Variants` property.

```

> {

Reviewer 2: ``` JavaScript
"MyFlag":
{   
    "Variants": [
        { 
            "Name": "Big", 
            "ConfigurationReference": "ShoppingCart:Big" 
        },  
        { 
            "Name": "Small", 
            "ConfigurationValue": {
                "Size": 300
            }
        } 
    ],
    "EnabledFor": [
        {
            "Name": "AlwaysOn"
        }
    ]
}
```
It would be more consistent to remove the "FeatureManagement" section and use JavaScript format to get the highlight.

If the feature is enabled, the feature manager will check the `User`, `Group`, and `Percentile` allocations in that order to allocate a variant for this feature. If the user being evaluated is named `Marsha`, in the group named `Ring1`, or the user happens to fall between the 0 and 10th percentile calculated with the given `Seed`, then the specified variant is returned for that allocation. In this case, all of these would return the `Big` variant. If none of these allocations match, the `DefaultWhenEnabled` variant is returned, which is `Small`.

> Allocation logic is similar to the [Microsoft.Targeting](./README.md#MicrosoftTargeting) feature filter, but there are some parameters that are present in targeting that aren't in allocation, and vice versa. Like targeting, allocation can only access user information once an implementation of `ITargetingContextAccessor` has been added to the service collection. Without this service, allocation can only return default variants if they're specified. The outcomes of targeting and allocation are not related.

Reviewer 2: Is this change lost?

Author: Talked with @jimmyca15 so I removed this, changed some titles to fit the sections better, and included targeting context earlier in the doc.

| Property | Description |
| ---------------- | ---------------- |
| `DefaultWhenDisabled` | Specifies which variant should be used when a variant is requested while the feature is considered disabled. |
| `DefaultWhenEnabled` | Specifies which variant should be used when a variant is requested while the feature is considered enabled and no other variant was assigned to the user. |
| `User` | Specifies a variant and a list of users to whom that variant should be assigned. | 
| `Group` | Specifies a variant and a list of groups the current user has to be in for that variant to be assigned. |
| `Percentile` | Specifies a variant and a percentage range the user's calculated percentage has to fit into for that variant to be assigned. |
| `Seed` | The value which percentage calculations for `Percentile` are based on. The percentage calculation for a specific user will be the same across all features if the same `Seed` value is used. If no `Seed` is specified, then a default seed is created based on the feature name. |

> In the above example, if the feature is not enabled, the feature manager will assign the variant allocated by `DefaultWhenDisabled` to the current user, which is `Small` in this case.

Reviewer: ```suggestion
In the above example, if the feature is not enabled, the feature manager will assign the variant marked as `DefaultWhenDisabled` to the current user, which is `Small` in this case.

```

> If the feature is enabled, the feature manager will check the `User`, `Group`, and `Percentile` allocations in that order to assign a variant. If the user being evaluated is named `Marsha`, in the group named `Ring1`, or the user happens to fall between the 0 and 10th percentile calculated with the given `Seed`, then the specified variant is assigned to the user. In this case, all of these would return the `Big` variant. If none of these allocations match, the user is assigned the `DefaultWhenEnabled` variant, which is `Small`.

Reviewer: ```suggestion
If the feature is enabled, the feature manager will check the `User`, `Group`, and `Percentile` allocations in that order to assign a variant. For this particular example, if the user being evaluated is named `Marsha`, in the group named `Ring1`, or the user happens to fall between the 0 and 10th percentile, then the specified variant is assigned to the user. In this case, all of these would return the `Big` variant. If none of these allocations match, the user is assigned the `DefaultWhenEnabled` variant, which is `Small`.

```

Reviewer: I would leave seed out of the explanation. In my opinion it raises the complexity too high.

Reviewer: Also, can we leave seed out of the flag example that covers overriding enabled/disabled with variants.

Reviewer 2: We can add an "**Advanced**" section to explain how seed works.

Author: I'm not really sure what to add in an Advanced section since the table above this part describes how `Seed` works and I feel like I might be reiterating that. Did you have anything in mind?

Reviewer 2: Oh, I didn't notice the table. Then we don't need a dedicated section for the seed usage.
